### DISCUSSION TITLE
Algorithm selection for collective communication operation

### DISCUSSION THREAD

**Discussion by zhuo121** [mpi4py/mpi4py#578] [Q&A]
I noticed that mpi4py provides some functions for collection communication, such as `MPI.Allreduce()`, `MPI.Ireduce_scatter()` and so on. However, collective communication operations may have different implementations. Take allreduce for example, there may be algorithms such as `ring` or `Recursive Doubling`. 
I would like to know what algorithm is used for collection communication operations in mpi4py? Can I choose the algorithm to use? Or do I have to implement the specific algorithms myself using `MPI.Send()` and `MPI.Recv()`?

**Comment by dalcinl** [comment]
mpi4py is a thin wrapper to a third-party MPI implementation. Therefore, anything related to tuning and algorithms is not under mpi4py's control, but on the backend MPI's shoulder. Different MPI implementations have different ways (e.g. config files, env variables) to select low level details such as collective algorithms. Check the documentation of your MPI implementation about how to select algorithms and perform any other runtime tuning.
